pkg/discovery/worker/compliance/affinity_helper.go

```diff
@@ -0,0 +1,302 @@
+package compliance
+
+import (
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	api "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/util"
+
+	"github.com/golang/glog"
+)
+
+// NOTE: implementations are copied and modified from scheduler/predicate
+
+//----------------------------------------- Node Affinity -------------------------------------------------------
+
+func matchesNodeSelector(pod *api.Pod, node *api.Node) bool {
+	// Check if node.Labels match pod.Spec.NodeSelector.
+	if len(pod.Spec.NodeSelector) > 0 {
+		selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
+		if !selector.Matches(labels.Set(node.Labels)) {
+			return false
+		}
+	}
+	return true
+}
+
+// The pod can only schedule onto nodes that satisfy requirements in both NodeAffinity and nodeSelector.
+func matchesNodeAffinity(pod *api.Pod, node *api.Node) bool {
+	nodeAffinityMatches := true
+	affinity := pod.Spec.Affinity
+	if affinity != nil && affinity.NodeAffinity != nil {
+		nodeAffinity := affinity.NodeAffinity
+		// if no required NodeAffinity requirements, will do no-op, means select all nodes.
+		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+			return true
+		}
+
+		// Match node selector for requiredDuringSchedulingIgnoredDuringExecution.
+		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
```

Comment:
there are three types:
[RequiredDuringSchedulingRequiredDuringExecution](https://github.com/kubernetes/client-go/blob/release-4.0/pkg/api/v1/types.go#L2257) - not implemented in upstream
[RequiredDuringSchedulingIgnoredDuringExecution](https://github.com/kubernetes/client-go/blob/release-4.0/pkg/api/v1/types.go#L2265) - implemented in Kubeturbo
[PreferredDuringSchedulingIgnoredDuringExecution](PreferredDuringSchedulingIgnoredDuringExecution) - Not handled in Kubeturbo for now